flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/StandaloneCompletedCheckpointStore.java

```diff
@@ -133,7 +133,13 @@ public void shutdown(JobStatus jobStatus, CheckpointsCleaner checkpointsCleaner)
 
             long lowestRetained = Long.MAX_VALUE;
             for (CompletedCheckpoint checkpoint : checkpoints) {
-                if (!checkpoint.discardOnShutdown(jobStatus)) {
+                if (checkpoint.shouldBeDiscardedOnShutdown(jobStatus)) {
+                    checkpoint.markAsDiscardedOnShutdown(jobStatus).discard();
```

Comment:
I don't fully sure about this place since the condition for OnShutdown checks twice in `shouldBeDiscardedOnShutdown` and `markAsDiscardedOnShutdown` it is possible instead of `markAsDiscardedOnShutdown` use `markAsDiscarded` but I would like to have consistency with `markAsDiscardedOnSubsume` and also I still want to encapsulate this logic somewhere to avoid the possible problem in the future.